Add support for the Meson build system

nielsdg opened the pull request:
See the [Meson website] for a full reference. To build, test and/or
install the folks library, you essentially need the following commands:

```
$ meson --prefix=/usr --sysconfdir=/etc _build
$ meson compile -C _build
$ meson test -C _build
$ meson install -C _build
```

[Meson website]: http://mesonbuild.com/

nielsdg committed:
Add support for the Meson build system

See the [Meson website] for a full reference. To build, test and/or
install the folks library, you essentially need the following commands:

```
$ meson --prefix=/usr --sysconfdir=/etc _build
$ meson compile -C _build
$ meson test -C _build
$ meson install -C _build
```

[Meson website]: http://mesonbuild.com/

BUG=https://github.com/ibus/ibus/pull/2358

mattst88 commented:
I would love to see this land. Autotools is just endless problems [like this](https://github.com/ibus/ibus/issues/2501).

thesamesam commented:
@fujiwarat Anything we could do to help get it over the line? Can rebase if needed.

fujiwarat commented:
Sorry for the late response.
I'm planning a big change with #2408 and I will ask this rebase after #2408 is landed.

jtojnar commented:
@fujiwarat Is it okay now? It looks like the change landed.

fujiwarat commented:
Yes, it would be great if someone could migrate the old patch to the git main.

fujiwarat committed:
github/workflows: Add Meson

Also enhance to show engine/simple.xml build flag.

BUG=https://github.com/ibus/ibus/pull/2358

fujiwarat commented:
/gemini review

fujiwarat commented on bindings/pygobject/meson.build:
```diff
@@ -0,0 +1,28 @@
+
+pyoverrides_tests_env = {
+  'PYTHONPATH': source_root / 'tests:@0@'.format(
+          python.get_variable('PYTHONPATH', '')),
```
Done.

fujiwarat commented on bindings/vala/meson.build:
```diff
@@ -0,0 +1,78 @@
+  libibus_emoji_dialog_valaflags = []
+  if get_option('emoji-dict')
+    libibus_emoji_dialog_valaflags += [ '-D', 'EMOJI_DICT' ]
+  endif
```
Done.

fujiwarat committed:
Update latest IBus and Meson

- Require Meson 1.5
- Require Python3
- GTK3, GTK4 should be build failures unless pkg-config files
- Use wayland.scan_xml() for Wayland protocol files
- Fix valist in gnome.genmarshal()
- Use gnome.compile_resources()
- Use gnome.generate_gir() for ibuswaylandim_gir, ibusimmodule_gir
- Use gnome.generate_vapi() for ibusimmodule_vapi, ibuswaylandim_vapi
- Make custom dependency of dbus_vapi, glibmacro_vapi for local VAPIs
- Delete python2_lib dependency
- Add systemd, libnotify, dbusmenu-gtk3, libevdev, libudev dependencies
- s/enable_vapigen/enable_vapi/
- Change Python environments for sysconfig.get_config_var()
  o s/'SCRIPTDIR'/'LIBDIR'/
  o s/'purelib'/'platlib'/
- s/meson_options.txt/meson.options/
- s/ibuskeynames.c/ibuskeys.c/
- s/X11_DATA_PREFIX/X11_LOCALEDATADIR/
- src/gencomposetable supports to save in a different directory
- src/emoji-parser outpus a null file instead of returning 99 since meson
  command in custom_target() needs to be succeeded.

Closes: #2358

fujiwarat committed:
meson: Refactor builds

- Use gi._overridesdir simply
- Fix Typelib dependency to build Vala files. Seems dependency() does not
  support Typelib files yet and use declare_dependency() instead.
- Add config.h.meson file for the header template
- Add summary() in meson.build
- Fix nested get_option('emoji-dict')

Closes: #2358

fujiwarat committed:
meson: Implement gtk-doc build

Closes: #2358

fujiwarat committed:
meson: Implement test build

meson: Enable the test build and build `memconf` with either
       enabled `memconf` or `test` option.
bindings/pygobject/meson: Enable the Python test with `runtest`.
bindings/pygobject/test-override-ibus: Add gi.require_versions().
bindings/pygobject/test-override-ibus: Add gi.require_versions().
bus/meson: Enable tests with `runtest`.
bus/test-matchrule: Support the TAP test.
bus/test-stress: Support the TAP test.
conf/memconf/meson: Separate the build and installation flags
engine/meson: Enable the Python test
src/ibuscomposetable: Replace g_print() with g_debug() for TAP comments.
src/tests/ibus-keypress: Prepend "# " for TAP comments.
src/tests/ibus-registry: Support the TAP test.
src/tests/ibus-util: Support the TAP test.
src/tests/meson:
 - Set protocl TAP.
 - Adjust the timeout.
 - Make a global `ibus_tests_env` dict but not environ since environment() +
   environment() is not a supported operation.
 - Append test suite strings by test conditions.
 - Use `runtest` driver.
 - Support a `xkb-latin-layouts` script test.
 - Set `is_parallel` flag to get focus events with the  expected time frame.
src/tests/runtest:
 - Expect the relative paths for Autotool and absolute paths for Meson.
 - Support Python tests.
 - Cease warnings with `glib-compile-schemas` command.
 - Enable subtests with TAP for Meson and set G_TEST_ROOT_PROCESS
   to consist test cases in Plan and ones in test results.
src/tests/xkb-latin-layouts: Support the TAP test.

Now `meson test` command is supported and you can disable some tests
with --no-suite option below.

% meson test -C builddir --list
% meson test -C builddir
% meson test -C builddir ibus-util
% meson test -C builddir --no-suite ibus:wm
% meson test -C builddir --wrapper "/bin/bash -x"

Closes: #2358

fujiwarat committed:
meson: dist command has new tests-exclude-suites option

Exclude test suites separated by comma for `meson dist`

% meson setup builddir -Dtests-exclude-suites=ibus:display
% meson dist -C builddir

Closes: https://github.com/mesonbuild/meson/issues/15733
Closes: #2358

fujiwarat committed:
github/workflows: Add Meson test

`meson test` command supports the `--no-suite` option to exclude some
tests but `meson dist` command does not support that option yet and
fails some tests at present because `meson dist` calls `meson test`
without any options internally.
I don't desire to use `meson dist --no-tests` option since it also
disables the build and whole the tests but wish to disable some tests
only.
The IBus CI now uses a special option
`-Dtests-exclude-suites=ibus:display` for `meson setup` command
so that `meson dist` can exclude some tests.

% meson setup builddir -Dtests-exclude-suites=ibus:display;

Closes: #2358

fujiwarat committed:
bus: Unset G_DISABLE_ASSERT for test builds

Now GTest is implemented in test-matchrule and test-stress for TAP
and G_DISABLE_ASSERT cannot be applied to the test builds.

Also delete the unnecessary GTK2_CFLAGS.

Closes: #2358

fujiwarat committed:
meson: Implement install-tests

Closes: #2358

fujiwarat committed:
meson: Fix misc issues

- po: Fix GETTEXT_PACKAGE domain.
- portal: Fix libexecdir.
- src/ibusenumtypes: Fix comments of @GType and delete build paths.
- src: Meson functions do not support the conditional installations yet
       and use meson.add_install_script() to delete 0 size dicts as the
       post installation.
- src/tests: Install uinput-replay to the test directory.

Closes: #2358

fujiwarat committed:
Replace setup/i18n.py and ui/gtk2/i18n.py

setup/i18n.py was a symlink of ui/gtk2/i18n.py but ui/gtk2 is deprecated
so replace the two files.

fujiwarat committed:
meson: Add meson-dist.py for additional dist logic

Closes: #2358

fujiwarat committed:
github/workflows: Support ubuntu:resolute

ubuntu:resolute is now LTS and it supports Meson 1.5
Wonder when I stop ubuntu:jammy.
